### DISCUSSION TITLE
jail.py error during restart of fail2ban  ("name 'noduplicates' is not _defined",)

### DISCUSSION THREAD

**Discussion by teoder** [fail2ban/fail2ban#3447] [Q&A] [answered]
When restarting fail2ban jails.py throws and error: _[1666]: ERROR   NOK: ("name 'noduplicates' is not _defined",)_

I assume it's cosmetic but would be nice to get rid of, anyone else seen this when running _"fail2ban-client restart"_?

**fail2ban log with DEBUG enabled:**
_root@fail2ban:/# tail -n 5 -f /config/log/fail2ban/fail2ban.log
  File "/usr/lib/python3.10/site-packages/fail2ban/server/jails.py", line 70, in add
   if noduplicates:
NameError: name 'noduplicates' is not defined_

I'm running Python 3.10.9 and fail2ban v1.0.2

**Comment by sebres** [comment, marked as answer]
> I assume it's cosmetic ...

It looks indeed like a careless mistake. And thanks for sharing the information. I'll fix this soon.

> would be nice to get rid of, anyone else seen this when running "fail2ban-client restart"?

I guess it'll happen only if you have duplicate jail sections somewhere in you configuration. 
Just I don't know how it is possible, since python merges duplicate section in config to single section (and fail2ban should do that in its includes too). 
Maybe something in your `/etc/fail2ban/jail.local` and in `/etc/fail2ban/jail.d/*.conf`? Just it is also very strange.

**Reply by teoder** [reply]
I had a bunch of unused jails in `/etc/fail2ban/jail.d/*` from the installation. Moved all of the to a subdir under jail.d and did a restart.
The error is gone! I'll try to figure out which one.

Thanks for pointing me in the right direction!

**Reply by sebres** [reply]
> I'll try to figure out which one.

Normally you'd be able to find it in the stream fail2ban sends to the fail2ban-server by start/restart. 
So you could try to dump it (temporary move the configs back to `jail.d`) and search for them in the output:
```bash
?sudo? fail2ban-client -d | grep "\['add',"
```
here you'd find all jails get added by start/restart.
